**Let's import relevant Libraries.**

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score,confusion_matrix,f1_score,classification_report
from collections import Counter
from imblearn.under_sampling import NearMiss
from imblearn.over_sampling import RandomOverSampler
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor

E:\Yash_Data_Part_2\Python\envs\py36-TF2.0\lib\importlib\_bootstrap.py:205: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)


In [17]:
df = pd.read_csv('creditcard.csv')

In [4]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [5]:
df.nunique()

Time      124592
V1        275663
V2        275663
V3        275663
V4        275663
V5        275663
V6        275663
V7        275663
V8        275663
V9        275663
V10       275663
V11       275663
V12       275663
V13       275663
V14       275663
V15       275663
V16       275663
V17       275663
V18       275663
V19       275663
V20       275663
V21       275663
V22       275663
V23       275663
V24       275663
V25       275663
V26       275663
V27       275663
V28       275663
Amount     32767
Class          2
dtype: int64

In [6]:
df.isna().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [7]:
df.describe(include='all')

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.165980e-15  3.416908e-16 -1.373150e-15  2.086869e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.490107e-15 -5.556467e-16  1.177556e-16 -2.406455e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.656562e-16 -3.444850e-16  2.578648e-16  4.471968e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.687098e-15 -3.666453e-16 -1.220404e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [18]:
df.shape

(284807, 31)

In [19]:
df.Class.value_counts()

0    284315
1       492
Name: Class, dtype: int64

**By now we know that we have total of 284315 records in this dataset and no missing values thankfully.**

**We have Time, Amount and Class features and other 28 features which have gone under PCA transformation for financial confidentiality.**

**Time feature shows time duration from the start.**

**Amount shows the amount of transaction.**

**Class shows whether the transaction was fraud or valid.**

**We can clearly see that out of 2.8 L transactions, only 400 were frauds. This indicates that we have a highly imbalanced dataset here.**

In [20]:
inputs = df.drop(columns=['Class'])
targets = df.Class

In [21]:
from sklearn.model_selection import train_test_split

In [22]:
x_train, x_test, y_train, y_test = train_test_split(inputs, targets, train_size=0.8, random_state=365)

In [26]:
y_train.value_counts()

0    227460
1       385
Name: Class, dtype: int64

# Supervised Learning

**Now that we have splitted our data into train and test data, let's discuss our plan for modelling.**

**We will use Random Forest Classification algorithm for problem at hand.** 

**Since the dataset is highly imbalanced, we will adopt 3 approaches to deal with it.**
1. **Class Weight**: We will give more weightage to class having less amount of records.


2. **Under Sampling**: We will reduce the number of records for class having huge number of records.


3. **Over Sampling**: We will insert duplicate records for class having less amount of records.


**It is notable here that we will only under/over sample train data and not the test data.**

In [27]:
class_weight = dict({0:1,1:100})

In [28]:
clf = RandomForestClassifier(class_weight=class_weight)
clf_us = RandomForestClassifier()
clf_os = RandomForestClassifier()

In [29]:
us = NearMiss(0.8)
os = RandomOverSampler(0.8)
x_train_us,y_train_us = us.fit_sample(x_train,y_train)
x_train_os, y_train_os = os.fit_sample(x_train,y_train)

E:\Yash_Data_Part_2\Python\envs\py36-TF2.0\lib\site-packages\imblearn\utils\_validation.py:638: FutureWarning: Pass sampling_strategy=0.8 as keyword args. From version 0.9 passing these as positional arguments will result in an error
  FutureWarning)


In [30]:
print("The number of classes before fit {}".format(Counter(y_train)))
print("The number of classes after UnderSampling {}".format(Counter(y_train_us)))
print("The number of classes after OverSampling {}".format(Counter(y_train_os)))

The number of classes before fit Counter({0: 227460, 1: 385})
The number of classes after UnderSampling Counter({0: 481, 1: 385})
The number of classes after OverSampling Counter({0: 227460, 1: 181968})


In [31]:
clf.fit(x_train,y_train)

RandomForestClassifier(class_weight={0: 1, 1: 100})

In [32]:
clf_us.fit(x_train_us,y_train_us)

RandomForestClassifier()

In [33]:
clf_os.fit(x_train_os,y_train_os)

RandomForestClassifier()

In [34]:
y_pred = clf.predict(x_test)

In [35]:
y_pred_us = clf_us.predict(x_test)

In [36]:
y_pred_os = clf_os.predict(x_test)

In [37]:
print('Confusion Matrix for Random Forest Classification with class weight')
print(confusion_matrix(y_test,y_pred))
print('Accuracy for Random Forest with class weight is ', accuracy_score(y_test,y_pred))
print('Confusion Matrix for Random Forest Classification with UnderSampling')
print(confusion_matrix(y_test,y_pred_us))
print('Accuracy for UnderSampling is ', accuracy_score(y_test,y_pred_us))
print('Confusion Matrix for Random Forest Classification with OverSampling')
print(confusion_matrix(y_test,y_pred_os))
print('Accuracy for OverSampling is ', accuracy_score(y_test,y_pred_os))

Confusion Matrix for Random Forest Classification with class weight
[[56852     3]
 [   26    81]]
Accuracy for Random Forest with class weight is  0.9994908886626171
Confusion Matrix for Random Forest Classification with UnderSampling
[[42953 13902]
 [    5   102]]
Accuracy for UnderSampling is  0.7558547803799024
Confusion Matrix for Random Forest Classification with OverSampling
[[56849     6]
 [   26    81]]
Accuracy for OverSampling is  0.9994382219725431


**Now let's understand what these measures indicate.**

**The accuracy score for our 3 approaches are 99%, 75% and 99% respectively.**

**Seems like our 1st and 3rd approaches have worked like a charm.**

**Here the confusion matrix comes into the picture** 

**By understanding geometry of confusion matrix, we can see that number of cases of predicting fraud transactions as valid are 26, 5 and 26 respectively.**

**This means our approach 2 has done wonderful job at predicting fraud transactions but has a poor performance in predicting valid transactions.**

**We can pass that for the case as we are looking for fraud transactions and so ironically the approach with minimum accuracy wins the case!!**

In [38]:
print('Classification Report for Random Forest Classification with class weight')
print(classification_report(y_test,y_pred))
print('Classification Report for Random Forest Classification with UnderSampling')
print(classification_report(y_test,y_pred_us))
print('Classification Report for Random Forest Classification with OverSampling')
print(classification_report(y_test,y_pred_os))

Classification Report for Random Forest Classification with class weight
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     56855
           1       0.96      0.76      0.85       107

    accuracy                           1.00     56962
   macro avg       0.98      0.88      0.92     56962
weighted avg       1.00      1.00      1.00     56962

Classification Report for Random Forest Classification with UnderSampling
              precision    recall  f1-score   support

           0       1.00      0.76      0.86     56855
           1       0.01      0.95      0.01       107

    accuracy                           0.76     56962
   macro avg       0.50      0.85      0.44     56962
weighted avg       1.00      0.76      0.86     56962

Classification Report for Random Forest Classification with OverSampling
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     56855
           1     

# Unsupervised Learning

**In such cases of highly unbalanced datasets, we can use unsupervised ML algorithms for anomaly/outlier detection.**

**Here we'll use 2 such algorithms.**

1. **Isolation Forest**
2. **Local Outlier Factor**

In [40]:
outlier_factor = df.Class.value_counts()[1]/df.Class.value_counts()[0]
outlier_factor

0.0017304750013189597

In [41]:
lof = LocalOutlierFactor()
y_pred_lof = lof.fit_predict(df.drop(columns=['Class']))
scores_prediction = lof.negative_outlier_factor_
scores_prediction

array([-1.50991882, -1.35540632, -1.98923719, ..., -1.82634219,
       -1.19589913, -1.50428432])

In [46]:
y_pred_lof[y_pred_lof==1]=0
y_pred_lof[y_pred_lof==-1]=1

In [47]:
isf = IsolationForest(contamination=outlier_factor)

In [48]:
isf.fit(df.drop(columns=['Class']))

IsolationForest(contamination=0.0017304750013189597)

In [49]:
y_pred_isf = isf.predict(df.drop(columns=['Class']))

In [50]:
y_pred_isf[y_pred_isf==1]=0
y_pred_isf[y_pred_isf==-1]=1

In [51]:
print('Confusion Matrix for Local Outlier Factor')
print(confusion_matrix(df.Class,y_pred_lof))
print('Accuracy for Local Outlier Factor is ', accuracy_score(df.Class,y_pred_lof))
print('Confusion Matrix for Isolation Forest')
print(confusion_matrix(df.Class,y_pred_isf))
print('Accuracy for Isolation Factor is ', accuracy_score(df.Class,y_pred_isf))

Confusion Matrix for Local Outlier Factor
[[280108   4207]
 [   393     99]]
Accuracy for Local Outlier Factor is  0.9838487115836339
Confusion Matrix for Isolation Forest
[[283959    356]
 [   355    137]]
Accuracy for Isolation Factor is  0.9975035725947747


In [52]:
print('Classification Report for Local Outlier Factor')
print(classification_report(df.Class,y_pred_lof))
print('Classification Report for Isolation Factor')
print(classification_report(df.Class,y_pred_isf))

Classification Report for Local Outlier Factor
              precision    recall  f1-score   support

           0       1.00      0.99      0.99    284315
           1       0.02      0.20      0.04       492

    accuracy                           0.98    284807
   macro avg       0.51      0.59      0.52    284807
weighted avg       1.00      0.98      0.99    284807

Classification Report for Isolation Factor
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    284315
           1       0.28      0.28      0.28       492

    accuracy                           1.00    284807
   macro avg       0.64      0.64      0.64    284807
weighted avg       1.00      1.00      1.00    284807



**Here you can see that accuracy in fraud prediction is considerable low when compared to supervised learning but we should also consider that in unsupervised learning everything our algorithm is seeing, it is seeing for the first time.**

**This can be improved by varying and using some parameters in these algorithms.**